### ISSUE TITLE
Configuring wildcard DNS is a major barrier for people running in a local VM

### ISSUE THREAD

**Issue by smcclos** [sandstorm-io/sandstorm#114]
Running Ubuntu 14.04 LTS.  Installed sandstorm fine, can configure to log onto github fine, installed 3 apps fine; wordpress; EtherCalc, and MediaGoblin.  But for all times when I want to start up a grain, get an error like this: 

Firefox can't find the server at 3xluodbuothxviaoa9u2.vu-sandstorm.localdomain.

where vu-sandstorm.localdomain is the name of my virtual machine.

**Comment by gemlog** [comment]
If you register a domain name (10-12$/year) with a registrar, then you could use ClouDNS for free. They make it easy to set up a cronjob to update your IP address, so you can easily use your own box at home with a dynamic IP. In addition, they have a simple web gui, so you can easily add an 'A' record for *.mydomain.tld creating a wildcard subdomain. Works fine for sub-sub-domains too e.g. *.subdomain.mydomain.tld -- no problem.

Both those features are free for up to three domain names. Naturally, you can create as many subdomains as you want for free for just one domain. 

Initially I had a typo in my .conf file where I'd fat-fingered WILDCARD_HOST, but kentonv kindly pointed that out to me and Boom! It just worked :-)

**Comment by bcg-didi** [comment]
I also had this issue and agree that it's a blocker for quick trying out in local LAN.

Workaround in my case (Ubuntu 12.04 client running the browser):
Create a file `/etc/NetworkManager/dnsmasq.d/dnsmasq.conf` with the content
`address=/.servername/192.168.0.100`
(whatever your server name and IP address are), then restart NetworkManager with
`sudo service network-manager restart`
I lost time trying with `/etc/dnsmasq.conf` which has no effect on Ubuntu.

**Comment by coyotebush** [comment]
It occurs to me that the install script could also check DNS (if `host`, `nslookup`, or such is available) and print a warning, to catch situations where people think they have everything set up but don't.

**Comment by mimoo** [comment]
Any how to? To setup this wildcard thing on debian?

**Comment by gemlog** [comment]
On Monday, April 13, 2015 08:02:14 David Wong wrote:
> Any how to? To setup this wildcard thing on debian?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/sandstorm-io/sandstorm/issues/114#issuecomment-92391567
I went through what I did here:
http://gemlog.ca/

**Comment by mimoo** [comment]
awesome I'll take a look at that

**Comment by JKGisMe** [comment]
+1 for this being confusing. I have a `my.group.company.com` domain I can play with. I can do any port there. But I can't do any sort of `*.my.group.company.com` and I was trying to use sandstorm bc I'm _not_ too familiar with server-y type stuff.

**Comment by gemlog** [comment]
On 2015-11-10 19:08, JKGisMe - notifications at github.com wrote:
>
> +1 for this being confusing. I have a |my.group.company.com| domain I
> can play with. I can do any port there. But I can't do any sort of
> |*.my.group.company.com| and I was trying to use sandstorm bc I'm
> /not/ too familiar with server-y type stuff.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sandstorm-io/sandstorm/issues/114#issuecomment-155648462>.
>

You might like to try their new ssl install. Sandstorm provides free ssl
certs (rotated weekly) and a free subdomain. For example mine is
gemlog.sandcats.io. You could be jkgisme.sandcats.io (or whatever
subdomain name you choose). The dynamic dns is automated and built-in
too. It's very cool. Please try that instead. It's quite painless
compared to the original install, which was already very painless -- if
you were familiar with server-y type stuff. Now you don't have to be.
Honest ;-)

I used Cloudns for making my first Sandstorm install and it had no
problem simply accepting a '*' in order to create a wildcard domain
name, so that's still on the table too for you.
paul

**Comment by JKGisMe** [comment]
I would try that for my personal projects but my company is kind of a stickler about using anything that isn't behind the firewall... basically i'd have to go through bureaucratic lvls of hell to get that approved... And then I'd have to figure out how to get past the firewall...

**Comment by gemlog** [comment]
On 2015-11-12 20:20, JKGisMe - notifications at github.com wrote:
>
> I would try that for my personal projects but my company is kind of a
> stickler about using anything that isn't behind the firewall...
> basically i'd have to go through bureaucratic lvls of hell to get that
> approved... And then I'd have to figure out how to get past the
> firewall...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sandstorm-io/sandstorm/issues/114#issuecomment-156318134>.
>
OK, well here is a thing I wrote up to do it without using sandcats.io
and their free ssl:
http://gemlog.ca

That little article and my previous post re the ssl install script is
all I know.

The new ssl install solves the twin problems of being able to have
trusted ssl certs and also the dynamic dns problem (the latter of which
you don't have at work, since you have a dedicated ip address already).
You'll also have to tweak whatever server is being used (nginx,
apache...) to direct sandstorm stuff to whatever internal box you are
installing on.

Truly, that's all I know about installing sandstorm and both ways worked
a treat. The peculiarities of your corporate network  environment and
which ports are open or mapped and how your internal network is laid
out... You have to be mostly on your own there, because you can't just
blab that info on the internet, obviously ;-)

If none of my help does it for you, I think you'll just have to get more
server-y I'm afraid (or better help :-) ). Really.

It may also be that it would be easier for you to host the one or two
server apps that you actually want by installing them on your target box
in the usual way behind the firewall? I don't know. Then you'll need to
know nginx/apache, mysql et al. and more networking good stuff...

I'm sorry. It is what it is JKGisMe. Computers are utterly simple; just
a bunch of switches that are either on or off. Then we add a zillion
layers of abstraction on top. You've simply bumped into some layers you
don't know yet perhaps, but you can learn them I'm sure.

For me, finding sandstorm was like finding
joomla/wordpress/moodle/whatever years ago after hand-bombing html and
js for years. Such a step up it's not even funny in terms of saving time
and easing discovery of new things.
paul

**Comment by kentonv** [comment]
Hey @JKGisMe, are you using Sandstorm for business purposes? Can I convince you to drop me an email at kenton at sandstorm.io to tell me about it? I'd love to understand your environment and see if we can find a solution that makes this easier for you.

**Comment by pumpadump** [comment]
Hi I would also love to try sandstorm on a local network behind a firewall without having to configure dnsmasq in every client. Is there now a possibility available to do so?

**Comment by kentonv** [comment]
@pumpadump Unfortunately, there's really no magic we can do that lets you escape from configuring DNS. However, instead of configuring DNS on every machine, I'd recommend configuring it centrally. There are basically two ways you could do this:

1. If you have an internal DNS server, or are willing to set one up, then configure a wildcard hostname for Sandstorm there.
2. You could actually use a public DNS service to configure a wildcard that maps to your internal IP. For people on your network, it will work, but for people outside your network, the site will go nowhere. You could, for example, register a new domain for $10 and then use Cloudflare (my employer) to handle DNS (for free). Then configure the wildcard to point to your internal IP address.

**Comment by ocdtrekkie** [comment]
It is probably worth mentioning that if you already own the domain, Cloudflare can still do this for you for free. :) I bring my own domain from elsewhere. In fact, it shouldn't even need to be a "new" domain, if you want to use a subdomain for Sandstorm and maybe use a prefix on your wildcard. (i.e. oasis.sandstorm.io)

**Comment by kentonv** [comment]
Right, I only suggested a new domain to avoid dependencies. But if there's an existing domain over whose DNS you have control, you can probably add a subdomain for Sandstorm, and a further wildcard subdomain on that, all pointing to the internal IP address.
